### PR TITLE
Arrays 

### PR DESCRIPTION
**Pull request by 20joshuaz**
So tired

### CODE REVIEW

File: lib/irgen.ml [hunk 1]
```diff
@@ -10,27 +10,26 @@ let translate (binds, sfuncs): L.llmodule =
   let f_t = L.double_type context 
   and i1_t = L.i1_type context 
   and i8_t = L.i8_type context 
+  and i32_t = L.i32_type context
```
Comment (Akash-Nayar) [inline lib/irgen.ml:13]: is this used for indexing?

File: lib/irgen.ml [hunk 2]
```diff
@@ -10,27 +10,26 @@ let translate (binds, sfuncs): L.llmodule =
...
-  let add_global (typ, name) = 
-    let init = 
-      match typ with
-          Number -> L.const_float (lltype_of_dtype typ) 0.0
-        | (Bool | Char) -> L.const_int (lltype_of_dtype typ) 0
-        | String -> L.const_stringz context ""
-        | _ -> raise (Failure "unimplemented")
-    in Hashtbl.add globals name (L.define_global name init mdl)
-  in List.iter add_global binds;
+  List.iter (fun (typ, name) -> Hashtbl.add globals name (L.define_global name (default_value typ) mdl)) binds;
```
Comment (Akash-Nayar) [inline lib/irgen.ml:32]: insanely big brain

File: lib/irgen.ml [hunk 3]
```diff
@@ -83,7 +82,32 @@ let translate (binds, sfuncs): L.llmodule =
         SNumberLit n -> L.const_float (lltype_of_dtype Number) n
       | SBoolLit b -> L.const_int (lltype_of_dtype Bool) (if b then 1 else 0)
       | SCharLit c -> L.const_int (lltype_of_dtype Char) (Char.code c)
-      | SStringLit s -> L.build_global_stringptr s "tmp" builder
+      | SStringLit s -> L.build_global_stringptr s "str" builder
+      | SArray arr -> 
+          let arr' = L.build_array_malloc (lltype_of_dtype (fst (List.hd arr))) (L.const_int i32_t (List.length arr)) "arr" builder in
+          List.iteri (
```
Comment (Akash-Nayar) [inline lib/irgen.ml:88]: bro knows the whole `List` library

File: lib/semant.ml [hunk 4]
```diff
@@ -61,6 +65,16 @@ let check (binds, funcs, stmts): sprogram =
...
       | CharLit chr -> (Char, SCharLit chr)
       | StringLit str -> (String, SStringLit str)
+      | ArrayLit arr -> 
+          if List.length arr = 0 then raise (Failure internal_err)
+          else 
+            let sarr = List.map check_expr arr in 
+            let typ = fst (List.hd sarr) in
+            List.iter (
+              fun item ->
+                if fst item <> typ then raise (Failure (inconsistent_array_err typ (fst item)))
```
Comment (Akash-Nayar) [inline lib/semant.ml:75]: lol

**Review by Akash-Nayar — APPROVED** [review]
literally the biggest brain PR yet

**Comment by erl-ang** [comment]
galaxy brain

File: lib/irgen.ml [hunk 5]
```diff
@@ -164,10 +191,11 @@ let translate (binds, sfuncs): L.llmodule =
           let pred_builder = L.builder_at_end context prd_bb in
           let bool_val = build_expr pred_builder prd in
       
-          let merge_bb = L.append_block context "merge" the_func in
+          let merge_bb = L.append_block context "merge_loop" the_func in
           ignore(L.build_cond_br bool_val block_bb merge_bb pred_builder);
           L.builder_at_end context merge_bb
       | SReturn sexp -> 
+          List.iter (fun arr -> ignore (L.build_free arr builder)) !malloced;
```
Comment (Akash-Nayar) [inline lib/irgen.ml:198]: simply commenting out this line fixes the bug. i guess we'll have to leave the `free`-ing up to the OS

File: lib/irgen.ml (hunk 1 above)
Comment (20joshuaz) [reply lib/irgen.ml:13]: yes I use i32's when I cast floats to ints (specifying size of malloc, specifying indexes, etc.). So there's gonna be a limit on the size of arrays but I think that number is big enough.

**Pull request merged**